sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/XdsProtocolTransformer.java

```diff
@@ -0,0 +1,171 @@
+            String serviceName, String clusterName) {
+        List<LocalityLbEndpoints> localityLbEndpointList = loadAssignment.getEndpointsList();
+        if (localityLbEndpointList == null) {
+            return new HashSet<>();
+        }
+        return localityLbEndpointList.stream()
+                .filter(Objects::nonNull)
+                .flatMap(localityLbEndpoints -> processLocalityLbEndpoints(localityLbEndpoints, serviceName,
+                        clusterName).stream())
+                .collect(Collectors.toSet());
+    }
+
+    private static Set<ServiceInstance> processLocalityLbEndpoints(LocalityLbEndpoints localityLbEndpoints,
+            String serviceName, String clusterName) {
+        List<LbEndpoint> lbEndpointsList = localityLbEndpoints.getLbEndpointsList();
+        if (lbEndpointsList == null) {
+            return new HashSet<>();
+        }
+        return lbEndpointsList.stream()
+                .filter(Objects::nonNull)
+                .map(lbEndpoint -> transformEndpoint2Instance(lbEndpoint, serviceName, clusterName,
+                        getInitializedMetadata(localityLbEndpoints)))
+                .collect(Collectors.toSet());
+    }
+
+    private static Map<String, String> getInitializedMetadata(LocalityLbEndpoints localityLbEndpoints) {
+        Map<String, String> metadata = new HashMap<>();
+        Locality locality = localityLbEndpoints.getLocality();
+        if (locality != null) {
+            metadata.put("region", locality.getRegion());
+            metadata.put("zone", locality.getZone());
+            metadata.put("sub_zone", locality.getSubZone());
+        }
+        return metadata;
+    }
+
+    private static ServiceInstance transformEndpoint2Instance(LbEndpoint endpoint, String serviceName,
+            String clusterName, Map<String, String> metadata) {
+        XdsServiceInstance instance = new XdsServiceInstance();
+        SocketAddress socketAddress = endpoint.getEndpoint().getAddress().getSocketAddress();
+        instance.setService(serviceName);
+        instance.setCluster(clusterName);
+        instance.setHost(socketAddress.getAddress());
+        instance.setPort(socketAddress.getPortValue());
+        endpoint.getMetadata().getFilterMetadataMap().values()
+                .forEach(struct -> struct.getFieldsMap()
+                        .forEach((key, value) -> metadata.put(key, value.getStringValue())));
+        instance.setMetadata(metadata);
+        if (HealthStatus.HEALTHY.equals(endpoint.getHealthStatus())
+                || HealthStatus.UNKNOWN.equals(endpoint.getHealthStatus())) {
+            instance.setHealthStatus(true);
+            return instance;
+        }
+        instance.setHealthStatus(false);
+        return instance;
+    }
+
+    private static Optional<String> getServiceNameFromCluster(String clusterName) {
+        if (clusterName == null) {
```

Comment:
You are advised to use StringUtils to determine whether the value is empty.

Reply:
done